### ISSUE TITLE
Create a global configuration store and inherit configuration for profiles

### ISSUE THREAD

**Issue by jorgemoralespou** [minishift/minishift#1517]
This is a RFE for adding the ability to have a global configuration store for minishift, where you can set configuration, addons, etc.. that would apply to any new created profile.

I would like all my instances to have 8 GB of memory and 2 GB of ram and always use virtualbox, and I want always to install the addon user by default.

```
$ minishift config set --global memory 8GB
$ minishift config set --global cpus 2
$ minishift config set --global vm-driver virtualbox
$ minishift addon enable --global admin-user
```

Then if I create a new profile(instance) I'll have that configuration applied:

```
$ minishift start --profile example
-- Creating vm using Virtualbox with 8GB ram and 2 cpus
```

If I want to provide some other configuration, I can do:

```
$ minishift config set --profile otherexample memory 4GB
$ minishift start --profile otherexample
-- Creating vm using Virtualbox with 4GB ram and 2 cpus 
```

or

```
$ minishift start --profile otherexample --memory 4GB
-- Creating vm using Virtualbox with 4GB ram and 2 cpus
$ minishift config list --profile otherexample
-- memory=4GB
-- cpu=2 (global)
-- vm-driver=virtualbox (global) 
```

In this way I can easily define what are my typical defaults, and what are my explicit non-defaults.

Then I can delete a profile safely, and create a new one, and will inherit the defaults if nothing else specified. You can probably do this currently via environment variables on your shell, but this is not intuitive, and a user would most likely not know where things are configured. This way provides consistency in that a user can do "minishift config list [--profile profile]" and see the configuration applied to the profile, either because it's global and inherited, or because the user has set it.

This should also be where cached images, oc clients, isos should be stored, in a global configuration space, and maybe only symlinked in the specific profile. This would help not bloat the system, and reduce startup speed of every new instance.

**Comment by hferentschik** [comment]
See also #1480

**Comment by hferentschik** [comment]
Also an alternative approach which could be a stepping  stone is the ability to copy a given profile. See https://github.com/minishift/minishift/issues/1480#issuecomment-335088965

Adding a _global_ option to the _config_ and _addon_ command leave several open questions which would need first discussing/resolution. A profile copy command could easily be added w/o breaking consistency. In the long run a _--global_ flag could exist in parallel to a profile copy command.

**Comment by jorgemoralespou** [comment]
Why adding a new command? copy/clone? What would it copy? What if I don't want to copy everything from one profile? Should I create "bare" profiles that I will not run just to hold the things I want to copy as "templates"?

**Comment by hferentschik** [comment]
> This should also be where cached images, oc clients, isos should be stored, in a global configuration space,

That's an independent issue really and yes, the aim is to share cached images, oc clients, isos

> and maybe only symlinked in the specific profile. 

no symlinks needed. Also on Windows that would not even work

**Comment by jorgemoralespou** [comment]
@hferentschik 

> How do we display global vs local config in commands like minishift config view? Just display the merged config

```
$ minishift config view [--profile otherexample]
-- memory=4GB
-- cpu=2 (global)
-- vm-driver=virtualbox (global) 
```

> Where do we store the shared config? allinstances.json? Should we in this case take the step and move the default profile into the profiles directory?

Yes. In a place where "minishift delete --profile xxx" would not destroy, and that by default if I just install minishift, I will have.

> minishift addon enable --global admin-user - what are the semantics here? What if this is not a default add-on. Does this command also copy the add-on or does it just enable it? In the latter, it would then get ignored when minishift start is called, since it would not be installed in the new profile (add-ons are not shared between profiles). If it is supposed to copy the add-on, the question becomes from where?

No, if it's not default addon, then a "install --global" would be needed.

**Comment by gbraad** [comment]
Global config is in line what my thinking when I worked on hostfolders. It was renamed to allinstances instead, but I still believe it is the better option.

Jumbling with copy and clones from other prifiles feels cumbersome and convoluted.

+1

**Comment by gbraad** [comment]
In hostfolders precendence is as follows

  * default
  * all instances (global) overrides
  * instance config override is final

In the case of hostfolders it is not displayed, but we can alter this.

**Comment by gbraad** [comment]
> Also on Windows that would not even work

Windows has junctions and these work similarly

**Comment by LalatenduMohanty** [comment]
I think `minishift profile copy` command makes sense as the first step and the global config can be a long run fix. Because IMO there will be two kind of users 
1. Users will copy one profile to another and then modify stuff. As it creates an impression that profiles are isolated from each other.
2. Users who will like to have global config which will be applicable to all profiles.

**Comment by LalatenduMohanty** [comment]
Created a separate issue https://github.com/minishift/minishift/issues/1649 for `profile copy/clone` feature

**Comment by jorgemoralespou** [comment]
@LalatenduMohanty where do we stand with this feature? IMHO this should be high priority, and I see it keep postponing.
When working with profiles, what most annoys me is the fact I need to repeat many configuration I always want for my profiles (mem, cpus, iso-url, image-cache,...). 

It's really annoying.

**Comment by LalatenduMohanty** [comment]
> @LalatenduMohanty where do we stand with this feature? IMHO this should be high priority, and I see it keep postponing.
> When working with profiles, what most annoys me is the fact I need to repeat many configuration I always want for my profiles (mem, cpus, iso-url, image-cache,...).
> 
> It's really annoying.

We are going to schedule it in the current sprint.

**Comment by LalatenduMohanty** [comment]
Resolved via https://github.com/minishift/minishift/commit/563b6f8e5fd9574d153e4d5c170de4d92aaf16e2 #2616
